glom/test/test_mutation.py

```diff
@@ -38,6 +38,8 @@ def r():
 def test_assign_spec_val():
     output = glom({'b': 'c'}, Assign('a', Spec('b')))
     assert output['a'] == output['b'] == 'c'
+    output = glom({'b': 'c'}, Assign('a', Or('d', 'b')))
```

Comment:
that's a good point; I was thinking about how to handle this and went through some ideas:

* capture constants into an {id(val): val} dict, to avoid re-walking the same data
* however "constant" to the glom eval means no spec, which isn't truly constant, so maybe copy.deepcopy
* for both this, and finding self-reference, a "compilation" step would be helpful; since this is a min-mode, we don't have the problem of passing through intermediate specs, so an explicit `Arg()` (used in all the other specs) could do constant folding as well as self-reference detection

Reply:
So, the job of CompileArg would be:
* replace python-constant trees (tuples, ints, floats, strings) with Val()
* replace constant trees with a spec Copy() or similar that does copy.deepcopy
* for others, apply the existing arg mode

```python
compile_arg({
   'constant': (1, 2, 3),
   'no-specs': [1, 2, 3],
   'arg-valuate': [T, T+1, T+2]
})
# this should yield
Arg({
   'constant': Val((1, 2, 3)),
   'no-specs': Copy([1, 2, 3]),
   'arg-valuate': [T, T+1, T+2],
})
```

Reply:
as long as self-references are contained to Val or Copy, no problem!

if they are in the path of a sub-spec, we can blow up in the spec constructor; since the point of ArgMode is to provide a mechanism of evaluating specs at all, this seems like a pure win

Reply:
okay, so my earlier idea opens a can of worms (see: https://github.com/mahmoud/glom/commit/e24cd300090b85fcf47bb8e97a4979f1bc9dd02a#diff-2c22213fb5b5d623d34e35549e442dc6f0221aba97b0ba261705282131006489R2565-R2636)

I think it would work, but its very complicated

I found a more straightforward way to make arg-mode recursion safe

(we can't make all specs in general recursion safe, because we don't know what the intention of user specs may be on repetition -- e.g. file.readline() or something shouldn't be memoized.  However, for arg-mode we know it can only ever recurse over itself, which means dict+list self recursion which we CAN guarantee aren't doing anything tricky like reading lines out of a file)